// @sap/ux-specification/dist/src/sync/v2/import/pages/analyticalListPage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../common/index");
const common_1 = require("../../../../specification/common");
const extensionLogger_1 = require("../../../../extensionLogger");
const i18n_1 = require("../../../../i18n/i18n");
const i18next_1 = __importDefault(require("i18next"));
const common_2 = require("../../../common");
const utils_1 = require("../../import/utils");
const listReport_1 = require("../pages/listReport");
const factory_1 = require("../../export/factory");
/**
 * Adds manifest settings to the config.json
 * @param {AnalyticalListPageConfig} analyticalListPageConfig - the configuration to be filled
 * @param {Manifest} manifest - manifest.json of the app
 * @param {object} jsonSchema - the app-specific schema
 * @param {MetadataInstanceInterface} factory - factory, for generating the access to reflect-metadata
 * @param {string} pageKey - key of the page in manifest
 * @param {ExtensionLogger} logger - instance for logging exceptions
 */
function addSettings(analyticalListPageConfig, manifest, jsonSchema, factory, pageKey, logger) {
    const tableType = index_1.determineTableType(manifest, pageKey);
    analyticalListPageConfig.table = factory.createInstance(common_1.PageType.AnalyticalListPage, `ALP${tableType}`, analyticalListPageConfig.table);
    index_1.transferSettingsOfObject(analyticalListPageConfig.table, manifest, jsonSchema['definitions'][`ALP${tableType}<LineItems>`], pageKey, logger);
    //Filterbar:
    analyticalListPageConfig.filterBar = factory.createInstance(common_1.PageType.AnalyticalListPage, 'AnalyticalListPageFilterBar', analyticalListPageConfig.filterBar);
    index_1.transferSettingsOfObject(analyticalListPageConfig.filterBar, manifest, jsonSchema['definitions']['AnalyticalListPageFilterBar'], pageKey, logger);
    //Chart settings:
    analyticalListPageConfig.chart = factory.createInstance(common_1.PageType.AnalyticalListPage, 'ChartSettings', analyticalListPageConfig.chart);
    index_1.transferSettingsOfObject(analyticalListPageConfig.chart, manifest, jsonSchema['definitions']['ChartSettings'], pageKey, logger);
    //KPI's:
    analyticalListPageConfig.keyPerformanceIndicators = {};
    const kpiCards = manifest[common_1.ManifestSection.generic]['pages'][pageKey].component.settings.keyPerformanceIndicators;
    if (kpiCards) {
        Object.keys(kpiCards).forEach((kpiKey) => {
            analyticalListPageConfig.keyPerformanceIndicators[kpiKey] = factory.createInstance(common_1.PageType.AnalyticalListPage, 'KPISettings');
            analyticalListPageConfig.keyPerformanceIndicators = factory.createInstance(common_1.PageType.AnalyticalListPage, 'KPISettings', analyticalListPageConfig.keyPerformanceIndicators);
            index_1.transferSettingsOfObject(analyticalListPageConfig.keyPerformanceIndicators[kpiKey], manifest, jsonSchema['definitions']['KPISettings'], pageKey, logger, kpiKey);
        });
    }
}
/**
 * Creates the configuration file content for an analytical list page V2
 * @param {object} manifest - manifest.json of the given application
 * @param {string[]} flex - list of all UI flexibility changes
 * @param appSchema - the application specific JSON schema
 * @param {FileData[]} fragments - list of fragment files and their content
 * @param {ExtensionLogger} logger - Logger class for logging messages
 *
 * @returns {AnalyticalListPageConfig} - the configuration (JSON)
 */
function createAnalyticalListPageConfig(manifest, flex, appSchema, fragments, logger) {
    // Initialize i18next
    i18n_1.initI18n();
    if (!manifest[common_1.ManifestSection.generic]) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOFE'),
            location: {
                path: common_2.MANIFESTPATH,
                range: [common_1.ManifestSection.generic]
            }
        });
        return;
    }
    const v2Pages = manifest[common_1.ManifestSection.generic].pages;
    if (!v2Pages) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOPAGES', { appId: manifest['sap.app']['id'] }),
            location: {
                path: common_2.MANIFESTPATH,
                range: [common_1.ManifestSection.generic]
            }
        });
        return;
    }
    const factory = new factory_1.MetadataInstanceFactoryV2();
    //Instantiate Page
    const config = factory.createPageInstance(common_1.PageType.AnalyticalListPage);
    const pageKeys = [];
    const v2Page = utils_1.findAnalyticalListPage(manifest[common_1.ManifestSection.generic].pages, pageKeys);
    if (!v2Page) {
        extensionLogger_1.log(logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOALP'),
            location: {
                path: common_2.MANIFESTPATH,
                range: [common_1.ManifestSection.generic]
            }
        });
        return;
    }
    addSettings(config, manifest, appSchema, factory, pageKeys[0], logger);
    flex.forEach((change) => {
        index_1.addFlex(config, change, appSchema);
    });
    listReport_1.addFragments(config, manifest, v2Page, common_1.PageType.AnalyticalListPage, fragments);
    return config;
}
exports.createAnalyticalListPageConfig = createAnalyticalListPageConfig;
