// @sap/ux-ui5-fe-mockserver-middleware/dist/router/request/odataV2Request.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODataV2Request = void 0;
var odataRequest_1 = require("./odataRequest");
var ODataV2Request = /** @class */ (function (_super) {
    __extends(ODataV2Request, _super);
    function ODataV2Request() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ODataV2Request.prototype.parseExpand = function (expandParameters) {
        var props = expandParameters.split(',');
        return props.reduce(function (reducer, property) {
            reducer.expand[property] = true;
            return reducer;
        }, { expand: {} });
    };
    // protected enrichElement(entity: string, element: object): void {
    //     const keyValues: string[] = [];
    //     const keys = this.metadata.getKeys(entity);
    //     keys.forEach((key) => {
    //         keyValues.push(`${key}='${element[key]}'`);
    //     });
    //     const uri = `${this.baseUrl}/${entity}(${keyValues.join(',')})`;
    //     element['__metadata'] = {
    //         id: uri,
    //         uri: uri,
    //         type: this.metadata.getEntitySet(this.query.entitySet).EntityType
    //     };
    // }
    ODataV2Request.prototype.createResponse = function (data, bCreated) {
        if (bCreated === void 0) { bCreated = false; }
        var response = _super.prototype.createResponse.call(this, data);
        response.setHeader('Content-Type', 'application/json');
        response.setHeader('dataserviceversion', '2.0');
        response.setHeader('sap-metadata-last-modified', 'Tue, 01 Oct 2019 15:02:44 GMT');
        response.setHeader('cache-control', 'no-store, no-cache');
        if (data && data.__metadata && data.__metadata.id) {
            response.setHeader('location', data.__metadata.id);
        }
        return response;
    };
    ODataV2Request.prototype.toJSON = function (data) {
        if (data.length !== undefined) {
            return JSON.stringify({
                d: {
                    __count: data.length + '',
                    results: data
                }
            });
        }
        else {
            return JSON.stringify({ d: data || {} });
        }
    };
    return ODataV2Request;
}(odataRequest_1.ODataRequest));
exports.ODataV2Request = ODataV2Request;
