// @sap/ux-ui5-fe-mockserver-middleware/dist/router/batch/content/appHttp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppHttp = void 0;
var AppHttp = /** @class */ (function () {
    /**
     * @param {String} url
     * @param {Object} headers
     * @param {Array} payload
     * @constructor
     */
    function AppHttp(url, headers, payload) {
        this.type = 'app';
        this.rawData = {
            url: url,
            headers: headers,
            payload: payload
        };
        this.contentId = undefined;
        this.request = undefined;
        this.response = undefined;
    }
    AppHttp.prototype.write = function (context, response) {
        response.write('Content-Type: application/http\r\n');
        response.write('Content-Transfer-Encoding: binary\r\n');
        if (this.contentId) {
            response.write('Content-ID: ' + this.contentId.id + '\r\n');
        }
        response.write('\r\n'); //end of header
        this.response.writeToBatchResponse(response, this.contentId);
    };
    return AppHttp;
}());
exports.AppHttp = AppHttp;
