// @sap/ux-ui5-fe-mockserver-middleware/dist/router/batch/content/batch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Batch = void 0;
var keyGenerator_1 = require("../keyGenerator");
var Batch = /** @class */ (function () {
    function Batch(type) {
        this.type = type;
        this.isChangeSet = type === 'changeset';
        if (this.isChangeSet) {
            this.changeSetErrorResponse = null;
        }
        this.parts = [];
    }
    Batch.prototype.write = function (context, response) {
        if (this.changeSetErrorResponse) {
            //if an error in the changeset occured then only the response of this error is returned
            response.statusCode = 200; //TODO check responce
            response.write('Content-Type: application/http\r\n');
            response.write('Content-Transfer-Encoding: binary\r\n');
            response.write('\r\n'); //end of header
            this.changeSetErrorResponse.writeToBatchResponse(response);
        }
        else {
            response.statusCode = 200;
            var boundary = 'batch_' + keyGenerator_1.createBoundary(context);
            if (this.isChangeSet) {
                response.write('Content-Type: multipart/mixed; boundary=' + boundary + '\r\n');
                response.write('\r\n'); //end of header
            }
            else {
                response.setHeader('Content-Type', 'multipart/mixed; boundary=' + boundary);
            }
            response.write('--' + boundary + '\r\n');
            for (var i = 0; i < this.parts.length; i++) {
                if (i !== 0) {
                    response.write('\r\n--' + boundary + '\r\n');
                }
                this.parts[i].write(context, response);
            }
            response.write('\r\n--' + boundary + '--\r\n');
        }
    };
    return Batch;
}());
exports.Batch = Batch;
