// @sap/ux-ui5-fe-mockserver-middleware/dist/router/batch/content/batchContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchContent = void 0;
function split(input) {
    var LF = '\n';
    var CRLF = '\r\n';
    var a = [];
    var pL = 0;
    var p1 = input.indexOf(CRLF, pL);
    var p2 = input.indexOf(LF, pL);
    while (p1 !== -1 || p2 !== -1) {
        if (p1 !== -1 && p1 <= p2) {
            a.push(input.substring(pL, p1));
            pL = p1 + 2;
        }
        else {
            a.push(input.substring(pL, p2));
            pL = p2 + 1;
        }
        p1 = input.indexOf(CRLF, pL);
        p2 = input.indexOf(LF, pL);
    }
    if (pL < input.length) {
        a.push(input.substring(pL));
    }
    return a;
}
var BatchContent = /** @class */ (function () {
    function BatchContent(data) {
        if (typeof data === 'string') {
            this.type = 0;
            this.stringData = data;
            this.stringSplit = split(data);
            this.pos = 0;
        }
        else {
            throw new Error('Only string allowed for batch parser');
        }
    }
    BatchContent.prototype.lookLine = function () {
        return this.stringSplit[this.pos];
    };
    BatchContent.prototype.readLine = function () {
        return this.stringSplit[this.pos++];
    };
    BatchContent.prototype.inc = function () {
        this.pos++;
    };
    return BatchContent;
}());
exports.BatchContent = BatchContent;
