// @sap/ux-ui5-fe-mockserver-middleware/dist/router/catalog.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.catalogService = void 0;
var router_1 = __importDefault(require("router"));
var fileLoader_1 = require("./utils/fileLoader");
function catalogService(config, version) {
    if (version === void 0) { version = 2; }
    var router = router_1.default();
    router.get('/ServiceCollection', function (req, res) {
        res.json({
            d: {
                results: [
                    {
                        ID: config.service.name
                    }
                ]
            }
        });
    });
    router.get('/ServiceCollection\\(*', function (req, res) {
        res.json({
            d: {
                results: [
                    {
                        TechnicalName: config.service.name,
                        Version: version
                    }
                ]
            }
        });
    });
    if (config.annotations) {
        router.get('/Annotations*', function (req, res) {
            fileLoader_1.loadFile(config.annotations.localPath).then(function (data) {
                res.contentType('application/xml').send(data);
            });
        });
    }
    return router;
}
exports.catalogService = catalogService;
