// @sap/ux-ui5-fe-mockserver-middleware/dist/router/batch/http/simpleHttpResponse.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponse = exports.SimpleResponse = void 0;
var http = __importStar(require("http"));
var SimpleResponse = /** @class */ (function () {
    function SimpleResponse() {
        this.statusCode = null;
        this.headers = {};
        this.data = '';
    }
    SimpleResponse.prototype.writeHead = function (code, headers) {
        this.statusCode = code;
        Object.assign(this.headers, headers);
    };
    SimpleResponse.prototype.setHeader = function (name, value) {
        this.headers[name] = value;
    };
    SimpleResponse.prototype.status = function (code) {
        this.statusCode = code;
        return this;
    };
    SimpleResponse.prototype.write = function (data) {
        this.data += data;
        return this;
    };
    SimpleResponse.prototype.end = function () {
        // nothing to do
    };
    SimpleResponse.prototype.writeToBatchResponse = function (response, contentId) {
        var value;
        var headers = this.headers;
        response.write('HTTP/1.1 ' + this.statusCode + ' ' + http.STATUS_CODES['' + this.statusCode] + '\r\n');
        Object.keys(headers).forEach(function (key) {
            value = headers[key];
            response.write(key + ': ' + value + '\r\n');
        });
        response.write('\r\n'); // end of header
        response.write(this.data);
        if (this.data.length > 0) {
            response.write('\r\n'); //as used in OData_Specification_V2.0 batch samples
        }
    };
    SimpleResponse.prototype.writeToResponse = function (res) {
        var _this = this;
        res.status(this.statusCode);
        Object.keys(this.headers).forEach(function (headerName) {
            res.header(headerName, _this.headers[headerName]);
        });
        if (this.data.length > 0) {
            res.send(JSON.parse(this.data));
        }
        else {
            res.send(this.data);
        }
    };
    return SimpleResponse;
}());
exports.SimpleResponse = SimpleResponse;
function createResponse() {
    return new SimpleResponse();
}
exports.createResponse = createResponse;
