// @sap/ux-ui5-fe-mockserver-middleware/dist/router/batch/http/simpleHttpRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRequestFromAppHttp = exports.createRequest = void 0;
var body_parser_1 = require("body-parser");
var clone = function (obj) {
    return JSON.parse(JSON.stringify(obj));
};
var createUrlData = function (request) {
    var protocol = request.connection && request.connection.encrypted ? 'https' : 'http';
    return {
        host: request.headers['x-forwarded-host'] || request.headers.host,
        proto: request.headers['x-forwarded-proto'] || request.protocol || protocol,
        path: request.headers['x-forwarded-path'] || request.url,
        forwarded: !!request.headers['x-forwarded-path'],
        baseUrl: request.baseUrl
    };
};
function createRequest(originalRequest, uriPrefix) {
    var request = {
        headers: {}
    };
    if (originalRequest) {
        request = {
            httpVersion: originalRequest.httpVersion,
            headers: clone(originalRequest.headers),
            method: originalRequest.method,
            url: originalRequest.url,
            urlData: createUrlData(originalRequest),
            baseUrl: uriPrefix || originalRequest.baseUrl || ''
        };
    }
    if (originalRequest.body) {
        request.body = originalRequest.body;
    }
    request.getBodyAsString = function (cb) {
        if (request.body) {
            if (request.body instanceof Buffer) {
                return cb(request.body.toString('utf8'));
            }
            else {
                return cb(request.body);
            }
        }
        else if (originalRequest.body) {
            return cb(originalRequest.body);
        }
        else {
            var parser = body_parser_1.raw({ type: '*/*' });
            parser(originalRequest, null, function next(err) {
                if (err) {
                    throw new Error('Error while parsing http content');
                }
                var body = originalRequest.body;
                return cb(body.toString('utf8'));
            });
        }
    };
    return request;
}
exports.createRequest = createRequest;
function createRequestFromAppHttp(appHttp, options) {
    var appData = appHttp.rawData;
    var s = appData.url.split(' ');
    if (s.length !== 3) {
        throw new Error('Invalid URI format inside batch request.');
    }
    var method = s[0];
    var uri = s[1];
    var version = s[2];
    var body = appData.payload.join('\r\n');
    if (uri.substr(0, 1) !== '/') {
        uri = '/' + uri;
    }
    var request = {
        httpVersion: version,
        headers: clone(appData.headers),
        method: method,
        url: uri,
        urlData: undefined,
        getBodyAsString: undefined,
        body: body
    };
    request.urlData = createUrlData(request);
    if (options) {
        Object.assign(request, options);
    }
    request.getBodyAsString = function (cb) {
        return cb(request.body);
    };
    return request;
}
exports.createRequestFromAppHttp = createRequestFromAppHttp;
